lib/pundit/policy_finder.rb

```diff
@@ -1,4 +1,46 @@
 module Pundit
+  class ScopeResolver
```

Comment:
It feels like a great improvement to extract a class to do this work, particularly since it separates out a responsibility from policy finder.

Since it's non-trivial, might there be value in testing its methods directly? Are there different combinations where the values of `resource_scope` and `policy` are various different types, which might be worth testing?

Also since this is a class we're calling directly, does it belong in its own named file?

Reply:
You're points are correct. Methods should be directly tested and this is the main reason I din't extract the class into a separate file. The assumptions behind this class is that it should be tested "directly" from `pundit_spec`. `PolicyFinder` does not have any unit tests as well so I kept the same behavior here. 

I agree to extract the class to a separate file. Let me gather all your comments and try to make some changes. Thanks.

Reply:
I actually think it would be fine to test through `pundit_spec` if you wanted to go that way -
- I think it's totally fine to have nested classes, particularly when they are unstable or where it's not clear where the seam should lie, since this makes them easier to change. The trade-off is often slightly more complicated or obtuse tests.

The thing is that in this case I'm not sure that all of the possible configurations of ScopeResolver are currently exercised by that test.

Probably easier to test in a separate file anyway 👍